### DISCUSSION TITLE
知乎阅读辅助增强

### DISCUSSION THREAD

**Discussion by Kenshin** [Kenshin/simpread#2861] [插件说明]
描述
---

> 优化知乎专栏与知乎问答页在阅读模式下的效果，如：**知乎问答页的描述**、**正常显示知乎公式（包括：阅读模式与导出 Markdown · HTML · PDF）** 等。

地址
---

[https://simpread.ksria.cn/plugins/details/IfgbkkZCOZ](https://simpread.ksria.cn/plugins/details/IfgbkkZCOZ)

适配效果
---
![image](https://github.com/Kenshin/simpread/assets/81074/df2fb1b4-8694-4193-8fad-a4fb0c95b32f)

选项
---

随便进入一个页面 e.g. https://sspai.com/post/71576 并进入 **阅读模式 → 右下角 → 动作 → 插件触发器**

按照下图所示即可查看此插件的选项页。（注意：如果你的界面没有插件触发器，请根据此 [解决方案](https://github.com/Kenshin/simpread/discussions/2342) 修复。

<img width="1141" alt="image" src="https://github.com/user-attachments/assets/9a874d28-6cbc-4724-91db-2b27687cfa3f">

功能
---

1. [在阅读模式正确显示知乎公式](https://github.com/Kenshin/simpread/discussions/2861#discussioncomment-3341890)

    <img width="510" alt="image" src="https://github.com/Kenshin/simpread/assets/81074/7cb737c7-b19d-4aee-978c-f59100bd953f">
    
    <img width="1251" alt="image" src="https://github.com/Kenshin/simpread/assets/81074/945f3d4f-3bfc-4c5b-a880-304be9de624b">

2. 导出 Markdown  / HTML / PDF 时可完美显示公式（需使用 [PDF  高级打印方案](https://github.com/Kenshin/simpread/discussions/3887)）

      <details><summary>导出 Markdown</summary>
          <img width="1440" alt="image" src="https://github.com/Kenshin/simpread/assets/81074/d6278f11-2b28-4594-983c-b39e4c993696">
      </details>
      <details><summary>导出 HTML</summary>
          <img width="1046" alt="image" src="https://github.com/Kenshin/simpread/assets/81074/7122a8ca-5777-4a75-b5fa-958b69222449">
      </details>
      <details><summary>在稍后读查看</summary>
          <img width="1842" alt="image" src="https://github.com/Kenshin/simpread/assets/81074/c54c8ac8-4a34-47ab-b965-8981438d27e1">
      </details>
      <details><summary>导出 PDF</summary>
          <img width="1355" alt="image" src="https://github.com/Kenshin/simpread/assets/81074/aab13e98-1ac3-4e7c-af60-264f3ffce167">
      </details>

3. 是否显示知乎问答页的问题描述

   <img width="496" alt="image" src="https://github.com/Kenshin/simpread/assets/81074/7999b140-28a7-4643-ae6f-c18c0fd3db37">

      <details><summary>知乎效果</summary>
          <img width="404" alt="image" src="https://github.com/user-attachments/assets/4ea5f30c-279c-4be8-850c-130b06a274ad">
      </details>
      <details><summary>开启后</summary>
          <img width="615" alt="image" src="https://github.com/user-attachments/assets/93c8afaf-739e-4884-a1e4-29b6515ac9a9">
      </details>

4. 去除知乎的外链跳转限制，并可选择外链的打开方式

     <img width="490" alt="image" src="https://github.com/Kenshin/simpread/assets/81074/b47aa898-bea7-4890-8a5e-86b854b2ee89">
     
5. 清除文内搜索和知乎直答的链接，仅保留【卡片链接】

      <details><summary>开启前</summary>
          <img width="1154" alt="image" src="https://github.com/user-attachments/assets/444e62e1-eb86-4306-85fb-6861f2545c3f">
      </details>
      <details><summary>开启后</summary>
          <img width="1107" alt="image" src="https://github.com/user-attachments/assets/ddf16837-6172-4998-8b2a-c795c8f23c79">
      </details>

7. [优化知乎问答页](https://github.com/Kenshin/simpread/discussions/2861#discussioncomment-6390836)，更适合导出

     <img width="488" alt="image" src="https://github.com/Kenshin/simpread/assets/81074/0f45937b-b282-4818-9328-9e01ec3775fd">

      <details><summary>导出效果</summary>
          <img width="793" alt="2024-10-10_19-14-26" src="https://github.com/user-attachments/assets/6a4df756-3821-40b3-896d-5d81e2856b92">
      </details>

8. 仅显示正文中的第一个答案

      <img width="483" alt="image" src="https://github.com/user-attachments/assets/83aa0789-0152-4cf2-b870-ec35b2a485c1">
      <details><summary>默认论坛页效果（开启前）</summary>
          <img width="1289" alt="image" src="https://github.com/user-attachments/assets/e11fd2f6-b735-4b99-b918-aff7ccfae5c5">
      </details>
      <details><summary>开启后</summary>
          <img width="1296" alt="image" src="https://github.com/user-attachments/assets/2790a864-7dde-48ec-9441-5087827701a7">
      </details>

9. 知乎问答页支持切换首个或全部答案（位置在：阅读模式右下角 → 触发器 → 动作 → 插件触发器）

    <img width="239" alt="image" src="https://github.com/user-attachments/assets/9290536c-d9aa-489e-a539-637c2d3c1d30">

安装与升级
---

- 没有【使用同步助手 · 自动同步】的话，可通过[此链接](https://simpread.ksria.cn/plugins/details/IfgbkkZCOZ) 安装。

- 已经【使用同步助手 · 自动同步】的话，请使用 [插件管理器](https://www.yuque.com/kenshin/simpread/fcw60u#ymS1C) 安装。

更新日志
---

### 2.1.1

- `Add` 切换首个或全部答案（位置在：阅读模式右下角 → 触发器 → 动作 → 插件触发器）

- `Add` 新版本验证

### 2.1.0

- `Add` 清理全部链接

- `Add` 仅显示第一个回答

- `Optimize` 自动清除知乎直答的链接中的 `svg` 内容

- `Optimize` 将知乎问答页的描述设置到 `sr-rd-desc` 节点

- `Optimize` 论坛页导出优化，去掉 `知乎 ID`

### 2.0.0

- `Add` [检查和添加 `zhihu.com` 到 延迟加载列表](https://github.com/Kenshin/simpread/discussions/2861#discussioncomment-3341808)

- `Add` [优化知乎问答页](https://github.com/Kenshin/simpread/discussions/2861#discussioncomment-6390836)，更适合导出

- `Optimize` 更新以及去掉了一些过时（错误的）链接和描述

- `Fix` 导出含有公式的页面为 HTML 无法显示公式的错误

- `Fix` 在稍后读无法查看知乎公式的错误

- `Fix` 导出含有公式的页面为 PDF 无法显示公式的错误（需使用 [PDF  高级打印方案](https://github.com/Kenshin/simpread/discussions/3887)）

### 1.1.1

- `Fix` 因 [知乎修改了公式规则](https://github.com/Kenshin/simpread/discussions/2861#discussioncomment-3341808) 此方案已失效，新版方案 [请看这里](https://github.com/Kenshin/simpread/discussions/2861#discussioncomment-3341890)。

### 1.1.0 

- 知乎问答页显示描述

- 去掉知乎内联外链

### 1.0.0

- 优化数学公式图片显示

- 进入阅读模式后自动将知乎公式图片转换为内联方式

注意
---

> 此功能仅运行在 `question.zhihu.com` `zhuanlan.zhihu.com` 域名下。

关联
---

> 关于完美适配知乎专栏与问答页的方式 [请看这里](https://github.com/Kenshin/simpread/discussions/2869)。

**Comment by gaesa** [comment]
~这个插件如果和"模仿Safari阅读模式样式"插件同时开启的话,还是会出现问题~

测试网址https://zhuanlan.zhihu.com/p/39281734

![字符排版出错](https://user-images.githubusercontent.com/71256557/133929051-30933d70-c030-47a5-a3b2-5664a7ec7871.png)

如图,"例二"后的排版明显有问题.但不开启"模仿Safari阅读模式样式"插件倒是没事.

### 造成此问题的解决方案

> 跳转阅读模式的版面宽度，即可解决，细节可以看 https://github.com/Kenshin/simpread/discussions/2861#discussioncomment-1355959

![image](https://user-images.githubusercontent.com/81074/133975751-bf826c24-e614-4ebd-9fd1-7d91f15e59a0.png)

**Reply by Kenshin** [reply]
不会的，上面的截图就是在这个插件基础上的。

另外请截图出有问题的地方，否则我没法分辨。

**Reply by Kenshin** [reply]
嗨，补个截图

![image](https://user-images.githubusercontent.com/81074/133952926-fd262a2f-53c7-488f-bd8d-7cf630b114c8.png)

**Reply by gaesa** [reply]
![屏幕截图 2021-09-20 153213](https://user-images.githubusercontent.com/71256557/133969102-d3ef6d6d-d50f-4271-ad73-b1aabc6d06da.png)
"例二"这个地方,不知道什么问题

**Reply by Kenshin** [reply]
> "例二"这个地方,不知道什么问题

这是在阅读模式下查看的吗？更换个不同的主题再看看。

**Reply by Kenshin** [reply]
也可能是版面布局的问题，你的浏览器的宽度过小吧？试着调整宽度到最大值。

**Reply by gaesa** [reply]
> 也可能是版面布局的问题，你的浏览器的宽度过小吧？试着调整宽度到最大值。

谢谢,在"样式"中把"版面宽度"跳到"82"就好了.

**Reply by gaesa** [reply]
不好意思,没看仔细,调大版面宽度只能解决部分问题,甚至在你的截图中也能看到问题.注意到,例三后三段的内容   "在上面三个例子中，我们有一个集合"   (就是你提供的截图的最下方的位置) 这个地方之后的字符奇怪地发生了错位,本来横向排列的字符跑到下一行去了,逗号与逗号之间产生了相当长的一段空格.

**Reply by Kenshin** [reply]
> 不好意思,没看仔细,调大版面宽度只能解决部分问题,甚至在你的截图中也能看到问题.注意到,例三后三段的内容 "在上面三个例子中，我们有一个集合" (就是你提供的截图的最下方的位置) 这个地方之后的字符奇怪地发生了错位,本来横向排列的字符跑到下一行去了,逗号与逗号之间产生了相当长的一段空格.

重新安装这个插件后再试试。（我调整了个地方，但需要重新安装）

**Reply by gaesa** [reply]
可以了,不受版面宽度影响,大致看下来没有什么大问题,不过颜色不一致的问题能否解决呢?

![屏幕截图 2021-09-20 184721](https://user-images.githubusercontent.com/71256557/133990560-f96a98d3-1b6c-48e0-a214-d829107b1847.png)

**Reply by Kenshin** [reply]
> 可以了,不受版面宽度影响,大致看下来没有什么大问题,不过颜色不一致的问题能否解决呢?

这个解决不了，这些本质上是图片。

**Reply by gaesa** [reply]
> > 可以了,不受版面宽度影响,大致看下来没有什么大问题,不过颜色不一致的问题能否解决呢?
> 
> 这个解决不了，这些本质上是图片。

但是知乎原始页面不会有这个问题

**Reply by Kenshin** [reply]
知乎的背景是白色的，这些公式也是白色的，但你使用的主题不是白色的。

**Reply by gaesa** [reply]
> 知乎的背景是白色的，这些公式也是白色的，但你使用的主题不是白色的。

好吧,谢谢告知

**Reply by Kenshin** [reply]
![image](https://user-images.githubusercontent.com/81074/133992357-bfe5372a-b364-4860-89c1-b6a72ce21082.png)

emmm 我看了下，这个「图片」是透明的，不应该出现你说的情况，上面是我的截图。

另外，你在 https://github.com/Kenshin/simpread/discussions/2861#discussioncomment-1356538 的截图也「明显」不对。

确定**重新安装插件和更新适配规则**了吗？

**Reply by gaesa** [reply]
插件是重装了,适配规则更新是怎么回事?我现在换一下主题色就没事

**Reply by gaesa** [reply]
> 插件是重装了,适配规则更新是怎么回事?我现在换一下主题色就没事

已经手动同步适配列表,不过还是需要修改主题色

**Reply by Kenshin** [reply]
我没办法给你截图看，你可以试试

进入这个页面的阅读模式，然后右下角控制面板 → 站点 → 拉到最下面 → 看看是不是有个叫「官方主适配源」→ 点一下

刷新页面，栏看是否有所改善。

**Reply by gaesa** [reply]
还是这样
![已启用](https://user-images.githubusercontent.com/71256557/133998170-c99e164c-b567-474b-b721-ae0d7e00b078.png)

**Comment by ccyousa** [comment]
深色模式下，公式都是暗色的：
![捕获](https://user-images.githubusercontent.com/34744812/147533690-24fc1f77-e9cc-4c60-83bc-fca23eb84c1d.PNG)
禁用后，就是原来的白色图片

**Reply by Kenshin** [reply]
确实如此，因为知乎公式都是图片，暂时没有好的解决方案。从另外一个角度说，也算不上是简悦的问题。  😂

**Comment by Theigrams** [comment]
能不能直接将知乎的公式图片转化为LaTeX，从而在markdown中以原生公式的形式显示呢？

理论上只需要做一个正则替换即可。

以下是参考方案：
- [知乎Markdown输入公式解决方案 - 知乎](https://zhuanlan.zhihu.com/p/69142198)
- [python爬取知乎页面的LaTeX公式_zorchp的博客-CSDN博客](https://zorchp.blog.csdn.net/article/details/110524808?spm=1001.2101.3001.6650.1&utm_medium=distribute.pc_relevant.none-task-blog-2%7Edefault%7ECTRLIST%7Edefault-1-110524808-blog-115938363.pc_relevant_eslanding_v3&depth_1-utm_source=distribute.pc_relevant.none-task-blog-2%7Edefault%7ECTRLIST%7Edefault-1-110524808-blog-115938363.pc_relevant_eslanding_v3&utm_relevant_index=2)

**Reply by Kenshin** [reply]
哈哈，可以考虑。

**Reply by Theigrams** [reply]
> 哈哈，可以考虑。

写了一个函数，可以把知乎公式图像的 `alt` 替换成 LaTeX 代码，希望能有所帮助

```javascript
(function() {
    "use strict";

    function latex_replace(img) {
        // get the latex code from data-formula
        var latex = img.getAttribute('data-formula');
        // add $ to the begin and end of the latex code
        latex = '$' + latex + '$';
        // replace the img alt with latex code
        img.setAttribute('alt', latex);
    }
    // elect all img element with eeimg="1"
    var imgs = document.querySelectorAll('img[eeimg="1"]');
    // replace the alt of img element
    for (var i = 0; i < imgs.length; i++) {
        latex_replace(imgs[i]);
    }
})();
```

**Reply by Kenshin** [reply]
哈哈，Cool ~

**Reply by OverflowCat** [reply]
今天在群里正好看到知乎公式的问题，想着来开个 issue。我之前写了一个用来复制 Markdown，可以区分 inline 和 block

```js
/**
 * decodeZhihuLaTeX
 * @param {string} src of the LaTeX <img>
 * @returns {string} decoded latex or empty string
 */
function decodeZhihuLaTeX(src) {
    if (src.startsWith("https://www.zhihu.com/equation?tex=")) {
        const encoded_latex = src.replace("https://www.zhihu.com/equation?tex=", "")
        return decodeURIComponent(encoded_latex.replace(/\+/g, '%20'))
    }
    return ""
}

document.querySelectorAll("img[src]").forEach(e => {
    const latex = decodeZhihuLaTeX(e.src)
    if (latex !== "") {
        // This is a math pic
        if (e?.parent?.chileNodes?.length != 1) {
            // This is an inline $…$
            e.outerHTML = `<span class="latex">\$${latex}\$</span>`
        } else {
            // This is a block $$…$$
            e.outerHTML = `<span class="latex">\$\$${latex}\$\$</span>`
        }
    }
})
```

**Comment by Kenshin** [comment]
# 关于知乎公式规则改变的说明

近期有用户反馈知乎公式的功能不好使了，我研究了下发现是因为知乎公式从原来的图片模式改变为类似 MathJax 方案导致。

![image](https://user-images.githubusercontent.com/81074/183274831-22ccc19b-2957-4b7c-a6ab-19e8c8145dbd.png)

因此旧方案已经失效，新方案 [请看这里](https://github.com/Kenshin/simpread/discussions/2861#discussioncomment-3341890)。

**Comment by Kenshin** [comment]
# 在阅读模式正确显示知乎公式

## 自动方案

这是 1.2.0 增加的新方案，可以将 `zhihu.com` 加入到 **延迟加载列表**。

<img width="965" alt="image" src="https://github.com/Kenshin/simpread/assets/81074/d8b759a1-c196-456f-916e-2f87553f71e0">

注意：加入后务必重启浏览器，确保生效（配置不会互相覆盖）。

### 视频操作

https://github.com/Kenshin/simpread/assets/81074/3ec95b1c-f059-4919-9f1e-8e28069ced8a

## 手动方案

安装此插件后，只需要将 `zhuanlan.zhihu.com` 或 `zhihu.com` 加入到延迟加载列表中即可。（注意：某些公式页面需要手动拉到最底部使其生效）

位置：**选项页 → 高级设定 → 词法分析引擎 → 延迟加载列表**，按照下图所示，**添加后刷新当前页面，看看是否生效。**

![image](https://user-images.githubusercontent.com/81074/183275479-a4610014-e070-46a2-86ad-e62470928af0.png)

### 视频演示

https://user-images.githubusercontent.com/81074/183275994-27759c17-1559-4225-8dc7-5f102fa8464d.mp4

## 引申

除了 `zhihu.com` 外，还可以添加具体的 URL e.g. https://zhuanlan.zhihu.com/p/413150417 这是延迟加载列表的功能，详细说明 [请看这里](http://ksria.com/simpread/docs/#/词法分析引擎?id=延迟加载)。

![image](https://user-images.githubusercontent.com/81074/183275512-be46d9d5-bcba-4465-b499-7a56bb71aa78.png)

**Reply by tigerlb** [reply]
Thanks!

**Reply by zbh9415** [reply]
![image](https://github.com/Kenshin/simpread/assets/109010562/37e539bc-a004-4e4f-959b-7da28bcb68bc)
![image](https://github.com/Kenshin/simpread/assets/109010562/2af4bb74-e48b-4670-b37b-691f71c99cab)
你好，想问问为什么装了插件后还是没办法用

**Reply by Kenshin** [reply]
@zbh9415 

需要提供 URL

**Reply by zbh9415** [reply]
这个吗？https://zhuanlan.zhihu.com/p/508561823 把这个放到哪吗

**Reply by zbh9415** [reply]
第一张图是我的稍后读里面打开的阅读模式，第二张图是我保存在obsidian里的md格式，我打打开网址然后按阅读模式后是能正常显示，但是保存到稍后读里面就是上面这两种图的情况。我也已经安装插件，并把zhuanlan.zhihu.com保存在延迟加载列表里面了。

**Reply by Kenshin** [reply]
@zbh9415

导入到 Obsidian 是没问题的，看下图

<img width="774" alt="image" src="https://github.com/Kenshin/simpread/assets/81074/62111077-1e06-44c9-ae6b-a7b28422a918">

这是导出的文本（下载后改名为 `.md` ） [一篇文章搞懂：向量自回归模型（VAR）理论、检验及评价.txt](https://github.com/Kenshin/simpread/files/11630904/VAR.txt)

> 第一张图是我的稍后读里面打开的阅读模式

知乎的这种模式在稍后读的显示是有问题的，只有导出的 Markdown 和阅读模式下的显示没问题。

**Reply by zbh9415** [reply]
谢谢，之前obsidian忘了切回阅读视图，但是我在稍后读和导出的html文件里看都一样不能正常显示。只能在obsidian看是这样吗？

**Reply by Kenshin** [reply]
@zbh9415

> 但是我在稍后读和导出的html文件里看都一样不能正常显示。

是的，我上一条已经说了... 😂

**Reply by zbh9415** [reply]
@Kenshin 好的非常感谢😊

**Comment by Kenshin** [comment]
# 优化知乎问答页

<img width="504" alt="image" src="https://github.com/Kenshin/simpread/assets/81074/7189d84d-5604-4d67-a13e-780a83463d66">

## 优化内容

- 去掉知乎用户的头像

- 将知乎用户 ID 转移到正文

- 使用分隔符分割

## 优化前

![image](https://github.com/Kenshin/simpread/assets/81074/a31d0240-3be9-4722-8a08-736c6233e3b5)

## 优化后

<img width="606" alt="image" src="https://github.com/Kenshin/simpread/assets/81074/933c7909-4ef2-4829-b6e5-7654512fa0f1">

## 导出为 Markdown

<img width="924" alt="image" src="https://github.com/Kenshin/simpread/assets/81074/9a68da95-03f5-4aee-a3b7-4207c1823dcb">

**Reply by Bioerjzy** [reply]
大佬，我打开按钮之后，进入阅读模式无法看到问答，只有专栏内容，请问要怎么解决？

**Reply by Kenshin** [reply]
问答页？还是专栏下面的评论？如果是后者的话，是正常的，简悦仅针对正文进行了优化。

**Reply by Bioerjzy** [reply]
好吧，是专栏评论，谢谢大佬！

**Reply by Kenshin** [reply]
okay 你可以使用 https://github.com/Kenshin/simpread/discussions/1554 来获取需要的评论（生成阅读模式），但可能不是你需要的方案。

**Comment by ccb00** [comment]
<img width="1315" height="782" alt="image" src="https://github.com/user-attachments/assets/3c9212df-d910-4838-b4ba-3edfe5236a7b" />
现在的阅读模式依旧对知乎文章的latex适配的不是很好，有什么能解决的办法吗

**Reply by Kenshin** [reply]
请给出 URL

**Reply by ccb00** [reply]
https://zhuanlan.zhihu.com/p/676818643

**Reply by Kenshin** [reply]
我这边没有任何问题，请仔细看这个插件的使用方法。

<img width="1246" height="581" alt="image" src="https://github.com/user-attachments/assets/ff216495-239d-47cc-a21a-f25630626112" />

<img width="1229" height="282" alt="image" src="https://github.com/user-attachments/assets/c9130307-93db-4204-a699-82dfe3532a7c" />

**Reply by ccb00** [reply]
我这里是一部分能识别，一部分识别不好，每次打开是不同的公式识别失败。BTW，插件使用方法在哪里看

**Reply by ccb00** [reply]
<img width="1102" height="570" alt="image" src="https://github.com/user-attachments/assets/12cbaa69-47d9-4335-b55a-5ac95ab5c1f4" />
 <img width="625" height="651" alt="image" src="https://github.com/user-attachments/assets/681ada7b-7d07-4668-b960-9e63fb55ffe3" />
我这里安装了插件了，但是插件触发器里没有对应的两个选项

**Reply by ccb00** [reply]
<img width="654" height="396" alt="image" src="https://github.com/user-attachments/assets/f70c5425-0523-4d81-93a6-7400dc488fa2" />

别的都基本解决了，只是如图这里，在小标题中的公式会错乱

**Reply by ccb00** [reply]
公式后面没有汉字的识别不出来

**Reply by Kenshin** [reply]
> 别的都基本解决了，只是如图这里，在小标题中的公式会错乱

<img width="341" height="213" alt="image" src="https://github.com/user-attachments/assets/17c6b58e-5fdc-4507-ad71-b82745e50be2" />


建议刷新页面后可以滚动条慢慢拉大最下面，让各种公式都先加载出来，然后再进入阅读模式。

**Reply by ccb00** [reply]
我是这样操作的，每次让他充分加载，但是这种后面没有汉字的公式所有的都会错乱。。。其他的，在段落中的都没什么问题

**Reply by Kenshin** [reply]
emmmm 

那就有点奇怪了，你可以考虑多测试几个不同的知乎页，看看是否可以稳定复现。

**Reply by ccb00** [reply]
<img width="806" height="491" alt="image" src="https://github.com/user-attachments/assets/16e906c2-06e4-4088-aadc-be418299ae3f" />
<img width="809" height="579" alt="image" src="https://github.com/user-attachments/assets/3313580e-6422-43e7-85ff-b0c4aa011da9" />
好像基本都出现在小标题中，主要是这种带角标的变量，显示有问题。不知道是知乎插件的问题？还是我其他设置可能有问题？

**Reply by Kenshin** [reply]
我目前没有任何问题，也没有看到其他用户报告此类问题。

我新建了一个 Issues 讨论这个情况，请在 https://github.com/Kenshin/simpread/issues/7221 讨论~
